### PR TITLE
fix: tune user DB pool observability

### PR DESCRIPTION
**Pull request by srstack**
## Summary
- split DB pool defaults by role and tune user/user_schema/meta defaults
- pass tenant_id through bounded user DB pool/probe metrics and health probe logs
- keep hot-path DB operation metrics aggregated by role/namespace to avoid tenant_id cardinality
- add tests for role-specific pool defaults, tenant-scoped DB pool/probe metrics, probe-state cleanup, and saturated bounded-pool probes

## Operational notes
- User DB pool/probe metrics intentionally carry tenant_id for tenant attribution, bounded by DRIVE9_POOL_MAX_TENANTS per pod. With the current defaults that is 1024 tenant pools * DRIVE9_USER_DB_MAX_OPEN_CONNS=4.
- Hot-path DB operation counters/histograms intentionally do not carry tenant_id. `drive9_db_operations_total` and `drive9_db_operation_duration_seconds_*` are role-level signals; use tenant-scoped DB pool/probe alerts and logs for tenant attribution.
- DRIVE9_DB_MAX_OPEN_CONNS and DRIVE9_DB_MAX_IDLE_CONNS are intentionally removed. Operators must migrate to role-specific settings: DRIVE9_META_DB_*, DRIVE9_USER_DB_*, and DRIVE9_USER_SCHEMA_DB_*.
- Bounded DB pool saturation is not treated as DB unavailability by the health probe; it is recorded as probe result="pool_saturated" and should be covered by pool wait/saturation alerts. This applies to meta, user, and user_schema pools.
- Alert rule follow-up: tidbcloud/runbooks#2629 updates user DB operation latency/error alerts to namespace-level aggregation and keeps tenant attribution on pool/probe alerts.

## Validation
- go test ./pkg/metrics
- go test ./pkg/server -run 'TestMetricsEndpoint|TestLeaderGatedWorkersOnLoseRacesOnLead'
- go test -p 1 ./pkg/metrics ./pkg/mysqlutil ./pkg/datastore ./pkg/tenant ./pkg/server ./cmd/drive9-server
- go test -p 1 ./pkg/metrics ./pkg/mysqlutil ./pkg/tenant ./cmd/drive9-server
- go test ./pkg/metrics ./pkg/mysqlutil ./pkg/tenant ./cmd/drive9-server

## Summary by CodeRabbit

* **New Features**
  * Added tenant-aware database opening and tenant-scoped Prometheus metrics for DB health and operations (including `tenant_id` when applicable).
  * Expanded metrics labeling and health probe behavior to track state per role/tenant.
  * Updated help/usage text with additional environment variables for tenant pool and per-role DB connection pool sizing.

* **Bug Fixes**
  * Improved DB recovery/unavailable logging to include tenant context when available.

* **Refactor**
  * Connection pool defaults are now role-aware, using role-specific environment variables and defaults.

### CODE REVIEW

File: pkg/metrics/db.go [hunk 1]
```diff
@@ -216,12 +251,14 @@ func (m *dbMetrics) probeOnce(ctx context.Context, timeout time.Duration, onChan
 
 	m.mu.Lock()
 	if m.probe == nil {
-		m.probe = map[string]dbProbeState{}
+		m.probe = map[dbMetricKey]dbProbeState{}
 	}
-	for role, agg := range results {
+	for keyString, agg := range results {
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/metrics/db.go]: Handled in the follow-up commits. probeOnce now rebuilds m.probe from the current registered DB set each cycle while comparing against the previous map for transitions, and TestDBHealthProbeDropsUnregisteredTenantState covers unregister cleanup.

File: pkg/metrics/db.go [hunk 2]
```diff
@@ -183,8 +218,8 @@ func (m *dbMetrics) probeOnce(ctx context.Context, timeout time.Duration, onChan
 			// out, PingContext blocks for a free slot and may hit `timeout`,
 			// recording the pool as unreachable when it is merely saturated. This is
 			// rarely hit because ApplyPoolDefaults leaves MaxOpenConns unlimited
-			// unless DRIVE9_DB_MAX_OPEN_CONNS is set; the pool-saturation alert
-			// covers that condition directly.
+			// unless the role-specific max-open env var is set; the pool-saturation
+			// alert covers that condition directly.
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/metrics/db.go]: Handled. The stale comment was updated to describe the role-specific non-zero defaults and the capacity planning formula, instead of saying MaxOpenConns remains unlimited.

File: cmd/drive9-server/main.go [hunk 3]
```diff
@@ -149,12 +149,17 @@ func main() {
 	// register. The startup ping above only proves reachability at boot.
 	probeInterval := time.Duration(envInt("DRIVE9_DB_HEALTH_PROBE_INTERVAL_SECONDS", 15)) * time.Second
 	probeTimeout := time.Duration(envInt("DRIVE9_DB_HEALTH_PROBE_TIMEOUT_SECONDS", 3)) * time.Second
-	metrics.StartDBHealthProbe(context.Background(), probeInterval, probeTimeout, func(role string, up bool, err error) {
+	metrics.StartDBHealthProbe(context.Background(), probeInterval, probeTimeout, func(info metrics.DBPoolInfo, up bool, err error) {
+		fields := []zap.Field{zap.String("role", info.Role)}
+		if info.TenantID != "" {
+			fields = append(fields, zap.String("tenant_id", info.TenantID))
+		}
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9-server/main.go:156]: Acknowledged. Keeping tenant-scoped DB health transitions at info/error intentionally for this rollout because the alert/log investigation path needs tenant_id. If volume is too high in production, we should tune aggregation/rate limits separately rather than dropping the tenant context in this PR.

File: pkg/mysqlutil/pool.go [hunk 4]
```diff
@@ -9,34 +9,99 @@ import (
...
+}
+
+func poolEnvInt(role, suffix string, def int) int {
+	if key := rolePoolEnvKey(role, suffix); key != "" {
+		if v, ok := lookupEnvInt(key); ok {
+			return v
+		}
+	}
+	return def
+}
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/mysqlutil/pool.go:82]: Intentional breaking change for this PR: the old global DRIVE9_DB_MAX_OPEN_CONNS / DRIVE9_DB_MAX_IDLE_CONNS knobs are removed rather than used as fallback, so role-specific limits cannot be silently mixed. The PR body now calls out the migration path to DRIVE9_META_DB_*, DRIVE9_USER_DB_*, and DRIVE9_USER_SCHEMA_DB_*.

File: pkg/mysqlutil/pool.go [hunk 5]
```diff
@@ -9,34 +9,99 @@ import (
...
+	}
+}
+
+func poolEnvInt(role, suffix string, def int) int {
+	if key := rolePoolEnvKey(role, suffix); key != "" {
+		if v, ok := lookupEnvInt(key); ok {
+			return v
+		}
+	}
+	return def
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/mysqlutil/pool.go:81]: Handled as an explicit migration note rather than a fallback. The old global env vars are intentionally not supported in this split because operators need to size meta, user, and user_schema pools separately. The PR body now documents the breaking change and replacement env vars.

File: pkg/mysqlutil/pool.go [hunk 6]
```diff
@@ -9,34 +9,99 @@ import (
...
 const (
-	defaultConnMaxLifetime = 5 * time.Minute
-	defaultConnMaxIdleTime = 45 * time.Second
+	defaultMetaConnMaxLifetime       = 5 * time.Minute
+	defaultMetaConnMaxIdleTime       = 45 * time.Second
+	defaultMetaMaxOpenConns          = 100
+	defaultMetaMaxIdleConns          = 20
+	defaultUserConnMaxLifetime       = 30 * time.Second
+	defaultUserConnMaxIdleTime       = 30 * time.Second
+	defaultUserMaxOpenConns          = 4
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/mysqlutil/pool.go:18]: Fixed in 33f513ff. Tenant pools that are already at MaxOpenConns with no idle connections now record probe result="pool_saturated" and are not counted as unreachable, so a busy tenant pool no longer flips drive9_db_up to 0. Added TestDBHealthProbeDoesNotMarkSaturatedTenantPoolDown.

[the remark this answers is not in the document]
Comment (srstack) [reply pkg/mysqlutil/pool.go:18]: Follow-up fixed in adc240cd: the saturation guard is no longer tenant-only. Any bounded pool, including meta and user_schema, records probe result="pool_saturated" without flipping drive9_db_up to 0. Added TestDBHealthProbeDoesNotMarkSaturatedMetaPoolDown for the control-plane case.

File: pkg/metrics/db_test.go [hunk 7]
```diff
@@ -123,3 +123,141 @@ func TestDBHealthProbeFlipsDriveDBUp(t *testing.T) {
...
+}
+
+func TestDBHealthProbeDropsUnregisteredTenantState(t *testing.T) {
+	const tenantID = "tenant-db-unregister-test"
+
+	healthy := &atomic.Bool{}
+	healthy.Store(false)
+	db := sql.OpenDB(fakeConnector{healthy: healthy})
+	db.SetMaxIdleConns(0)
+	t.Cleanup(func() { _ = db.Close() })
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/metrics/db_test.go]: Fixed in 3d66274f. The tenant DB metrics/probe tests now unregister their DB and run one probe cycle in t.Cleanup before closing, matching the meta-pool cleanup and preventing stale globalDB.probe state from leaking into later tests. Verified with go test ./pkg/metrics.

**Review by qiffang — CHANGES_REQUESTED** [review]
## PR #650 — Per-role DB pool defaults + tenant-level DB metrics

**Verdict: REVISE — 2 blockers, 1 nit**

**Scope**: 1167 lines across 12 files. Three changes bundled:
1. Role-specific DB pool defaults — meta (100/20), user (4/1), user_schema (8/0) with per-role env var overrides
2. Tenant-level DB metrics — `tenant_id` label on `drive9_db_*` Prometheus series for user pools
3. Saturated pool skip — health probe skips PingContext for tenant pools at MaxOpenConns to avoid blocking

---

### B1 — `ApplyPoolDefaults` signature change is a breaking API (BLOCKER)

`ApplyPoolDefaults(db *sql.DB)` → `ApplyPoolDefaults(db *sql.DB, role string)`. This is an exported function in `pkg/mysqlutil`. Any external caller or internal caller not updated in this PR will fail to compile.

The old env vars `DRIVE9_DB_MAX_OPEN_CONNS` / `DRIVE9_DB_MAX_IDLE_CONNS` are silently removed — existing deployments that set them will have no effect, and the new defaults (meta: 100 open, user: 4 open) will silently apply. This is a breaking behavior change with no migration path.

**Why this matters**: Operators who tuned `DRIVE9_DB_MAX_OPEN_CONNS=50` for their cluster sizing will silently get meta=100 and user=4 after upgrading. If their TiDB `max_connections` was sized for 50 per pod, they'll hit connection limits.

Must either:
- Continue reading the old env vars as fallback
- Log a deprecation warning when old vars are set but new ones aren't

---

### B2 — Cardinality explosion from `tenant_id` on DB metrics (BLOCKER)

Every tenant pool now emits `tenant_id` as a Prometheus label on `drive9_db_up`, `drive9_db_pool_registered`, `drive9_db_pool_connections`, `drive9_db_pool_wait_count_total`, `drive9_db_pool_closes_total`, `drive9_db_unreachable_pools`, `drive9_db_probe_duration_seconds`, AND per-operation `drive9_db_operations_total`.

With `defaultTenantPoolMaxTenants = 1024` per pod, this creates up to **1024 unique label combinations per metric**. For `drive9_db_pool_connections` with 4 state values × 1024 tenants = **4096 series just from that one metric**. For `drive9_db_operations_total` with ~7 operations × 2 results × 1024 tenants = **~14,336 series**.

Prometheus scrape, storage, and query cost scale linearly with series count. This will significantly impact monitoring infrastructure.

The `drive9_db_operations_total` instrumentation with `tenant_id` is especially dangerous because it's called on **every DB operation** (every query, exec, prepare, commit, rollback) — not just per-probe-cycle. The cardinality is unbounded by the number of distinct tenants that have ever been active.

**Why this matters**: This is a known anti-pattern. Tenant-level DB operation metrics should use a separate high-cardinality pipeline (logs, traces) rather than Prometheus labels.

Suggestion: keep `tenant_id` on pool-level gauges (these are bounded by `maxTenants`) but **remove it from per-operation counters/histograms** (`drive9_db_operations_total`, `drive9_db_operation_duration_seconds_bucket`).

---

### N1 — `isTenantPoolSaturated` silently skips probe — no unreachable count, no error aggregation (nit)

When `isTenantPoolSaturated` returns true, the goroutine returns immediately without updating `results[key]`. The pool was pre-counted in `results[key].total` but is neither reachable nor unreachable — it's silently omitted from the aggregate.

The probe still reports `up=1` (since `unreachable == 0`), which is correct for the intended behavior. But `drive9_db_pool_registered` count includes it while the probe effectively ignored it. A cleaner approach would be to not count it in `total` either, or add a separate `saturated` counter.

---

### Checklist

| Check | Result |
|---|---|
| All inputs handled? | **NO** — old env vars silently dropped (B1) |
| Edge cases covered? | Partial — saturated pool skip logic works but accounting is off (N1) |
| Failure behavior defined? | YES |
| Race/state issues? | NO — probe state correctly rebuilt each cycle |
| Tests complete? | YES for new features; missing test for old env var migration (B1) |
| Cardinality impact? | **UNBOUNDED** — tenant_id on per-operation metrics (B2) |

**B1 and B2 must be resolved before GREEN.**

**Comment by srstack** [comment]
Addressed the review as follows:

- B1: not changing the old global env behavior. `DRIVE9_DB_MAX_OPEN_CONNS` / `DRIVE9_DB_MAX_IDLE_CONNS` are intentionally removed in this split so operators must size meta/user/user_schema separately. The PR body calls this out as a breaking migration to `DRIVE9_META_DB_*`, `DRIVE9_USER_DB_*`, and `DRIVE9_USER_SCHEMA_DB_*`. I also kept `ApplyPoolDefaults(db, role)` as a required role API; we do not expect external or missed internal callers using `ApplyPoolDefaults(db)`.
- B2: fixed in `9299ac2d`. Removed `RecordTenantDBOperation`; DB operation counters/histograms now record only `role`, `operation`, and `result`. `tenant_id` remains only on bounded pool/probe metrics. Updated the test to assert DB operation metrics do not carry `tenant_id`.
- Alerts: opened tidbcloud/runbooks#2629 to align alert rules with B2. User DB operation latency/error alerts are now namespace-level; tenant attribution stays on pool/probe alerts and logs.
- N1: leaving saturated pools as `result="pool_saturated"` and not counting them unreachable by design. Pool wait/saturation alerts cover that state; `drive9_db_up` is reserved for actual probe failures rather than pool-capacity exhaustion.

Validation:
- `go test ./pkg/metrics ./pkg/mysqlutil ./pkg/tenant ./cmd/drive9-server`

**Review by qiffang — CHANGES_REQUESTED** [review]
## Re-review: cardinality blocker resolved, env var migration still open

### Resolved: cardinality explosion (Blocker 2)

`observeDBOperation` now discards `tenantID` (`_` parameter). Per-operation metrics (`drive9_db_operations_total`, `drive9_db_operation_duration_seconds`) carry only `role`. `tenant_id` only appears on pool-level metrics, bounded by `DRIVE9_POOL_MAX_TENANTS` (1024). ~10K series — manageable.

### Still open: breaking env var migration (Blocker 1)

Old `DRIVE9_DB_MAX_OPEN_CONNS` / `DRIVE9_DB_MAX_IDLE_CONNS` are removed with no fallback. Operators with these set will silently lose their overrides. The new meta default (100 open) may exceed what an operator had configured (e.g. 50) to stay within TiDB `max_connections`.

**Suggested fix**: add a fallback check for the old env vars with a deprecation log warning, or at minimum document the required migration in the PR description and release notes.

### Non-blocking observations

1. `isDBPoolSaturated` skips ping for saturated pools — if ALL pools for a key are saturated, a genuine DB outage is masked. Acceptable (pool-saturation metrics cover this).
2. `MaxTenants` default bump 128→1024: with user pools at 4 open conns, peak = 4096 connections per pod. Verify this fits within TiDB `max_connections` budget.

**Review by qiffang — APPROVED** [review]
## Re-review: GREEN

Both blockers resolved:

1. **Breaking env var migration**: PR body now explicitly documents that DRIVE9_DB_MAX_OPEN_CONNS / DRIVE9_DB_MAX_IDLE_CONNS are intentionally removed, with migration path to role-specific vars. Reasonable choice — global fallback would silently apply the wrong limits to per-role pools.

2. **Cardinality explosion**: `observeDBOperation` discards tenantID (`_` parameter). Per-operation metrics remain role-level. Pool-level metrics carry tenant_id, bounded by DRIVE9_POOL_MAX_TENANTS (1024).

3. **Saturated pool probe**: Fixed — bounded pools at capacity record `result="pool_saturated"` without flipping `drive9_db_up`. Covers meta, user, and user_schema.

4. **Test cleanup**: globalDB.probe leak fixed per CodeRabbit suggestion.

Approved.
